lib/collection/src/shards/replica_set/execute_read_operation.rs

```diff
@@ -113,7 +113,19 @@ impl ShardReplicaSet {
     where
         F: Fn(&(dyn ShardOperation + Send + Sync)) -> BoxFuture<'_, CollectionResult<Res>>,
     {
-        let local = self.local.read().await;
+        let local = match self.local.try_read() {
+            Ok(local) => local,
+            Err(_) => {
+                // Reject read request, if
+                // - `local` shard is not readily available
+                // - and partial snapshot recovery is in progress
+                self.check_partial_snapshot_recovery_lock()?;
```

Comment:
Technically it is possible that `snapshot_recovery_lock` will be acquired right after our check, and we end up waiting for the  shart recovery here